File: packages/capacitor-actito-scannables/android/src/main/java/com/actito/scannables/capacitor/ActitoScannablesPlugin.kt
package com.actito.scannables.capacitor

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.actito.Actito
import com.actito.ActitoCallback
import com.actito.scannables.ActitoScannables
import com.actito.scannables.ktx.scannables
import com.actito.scannables.models.ActitoScannable

@CapacitorPlugin(name = "ActitoScannablesPlugin")
public class ActitoScannablesPlugin : Plugin(), ActitoScannables.ScannableSessionListener {
    override fun load() {
        logger.hasDebugLoggingEnabled = Actito.options?.debugLoggingEnabled ?: false

        EventBroker.setup(this::notifyListeners)

        Actito.scannables().removeListener(this)
        Actito.scannables().addListener(this)
    }

    @PluginMethod
    public fun canStartNfcScannableSession(call: PluginCall) {
        call.resolve(JSObject().apply {
            put("result", Actito.scannables().canStartNfcScannableSession)
        })
    }

    @PluginMethod
    public fun startScannableSession(call: PluginCall) {
        val activity = activity ?: run {
            call.reject("Cannot start a scannable session without an activity.")
            return
        }

        Actito.scannables().startScannableSession(activity)
        call.resolve()
    }

    @PluginMethod
    public fun startNfcScannableSession(call: PluginCall) {
        val activity = activity ?: run {
            call.reject("Cannot start a scannable session without an activity.")
            return
        }

        Actito.scannables().startNfcScannableSession(activity)
        call.resolve()
    }

    @PluginMethod
    public fun startQrCodeScannableSession(call: PluginCall) {
        val activity = activity ?: run {
            call.reject("Cannot start a scannable session without an activity.")
            return
        }

        Actito.scannables().startQrCodeScannableSession(activity)
        call.resolve()
    }

    @PluginMethod
    public fun fetch(call: PluginCall) {
        val tag = call.getString("tag") ?: run {
            call.reject("Missing 'tag' parameter.")
            return
        }

        Actito.scannables().fetch(tag, object : ActitoCallback<ActitoScannable> {
            override fun onSuccess(result: ActitoScannable) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", result.toJson())
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    // region ActitoScannables.ScannableSessionListener

    override fun onScannableDetected(scannable: ActitoScannable) {
        try {
            EventBroker.dispatchEvent("scannable_detected", scannable.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the scannable_detected event.", e)
        }
    }

    override fun onScannableSessionError(error: Exception) {
        try {
            EventBroker.dispatchEvent("scannable_session_failed", JSObject().apply {
                put("error", error.localizedMessage)
            })
        } catch (e: Exception) {
            logger.error("Failed to emit the scannable_session_failed event.", e)
        }
    }

    // endregion
}

File: packages/capacitor-actito-scannables/ios/Sources/ActitoScannablesPlugin/ActitoScannablesPlugin.swift
import ActitoKit
import ActitoScannablesKit
import Capacitor
import Foundation

@objc(ActitoScannablesPlugin)
public class ActitoScannablesPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ActitoScannablesPlugin" 
    public let jsName = "ActitoScannablesPlugin" 
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "canStartNfcScannableSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startScannableSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startNfcScannableSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startQrCodeScannableSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "fetch", returnType: CAPPluginReturnPromise),
    ] 
    private var rootViewController: UIViewController? {
        get {
            UIApplication.shared.delegate?.window??.rootViewController
        }
    }

    public override func load() {
        addApplicationLaunchListener()

        EventBroker.instance.setup { self.notifyListeners($0, data: $1) }
        Actito.shared.scannables().delegate = self
    }

    @objc func canStartNfcScannableSession(_ call: CAPPluginCall) {
        call.resolve([
            "result": Actito.shared.scannables().canStartNfcScannableSession
        ])
    }

    @objc func startScannableSession(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let rootViewController = self.rootViewController else {
                call.reject("Cannot start a scannable session with a nil root view controller.")
                return
            }

            Actito.shared.scannables().startScannableSession(controller: rootViewController)
            call.resolve()
        }
    }

    @objc func startNfcScannableSession(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Actito.shared.scannables().startNfcScannableSession()
            call.resolve()
        }
    }

    @objc func startQrCodeScannableSession(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let rootViewController = self.rootViewController else {
                call.reject("Cannot start a scannable session with a nil root view controller.")
                return
            }

            Actito.shared.scannables().startQrCodeScannableSession(controller: rootViewController, modal: true)
            call.resolve()
        }
    }

    @objc func fetch(_ call: CAPPluginCall) {
        guard let tag = call.getString("tag") else {
            call.reject("Missing 'tag' parameter.")
            return
        }

        Actito.shared.scannables().fetch(tag: tag) { result in
            switch result {
            case let .success(scannable):
                do {
                    call.resolve([
                        "result": try scannable.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

}

extension ActitoScannablesPlugin: ActitoScannablesDelegate {
    public func actito(_ actitoScannables: ActitoScannables, didDetectScannable scannable: ActitoScannable) {
        do {
            EventBroker.instance.dispatchEvent("scannable_detected", data: try scannable.toJson())
        } catch {
            logger.error("Failed to emit the scannable_detected event.", error: error)
        }
    }

    public func actito(_ actitoScannables: ActitoScannables, didInvalidateScannerSession error: Error) {
        EventBroker.instance.dispatchEvent("scannable_session_failed", data: [
            "error": error.localizedDescription
        ])
    }
}

extension ActitoScannablesPlugin {
    private func addApplicationLaunchListener() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(didFinishLaunching),
            name: UIApplication.didFinishLaunchingNotification,
            object: nil
        )
    }

    private func removeApplicationLaunchListener() {
        NotificationCenter.default.removeObserver(
            self,
            name: UIApplication.didFinishLaunchingNotification,
            object: nil
        )
    }

    @objc private func didFinishLaunching() {
        removeApplicationLaunchListener()

        logger.hasDebugLoggingEnabled = Actito.shared.options?.debugLoggingEnabled ?? false
    }
}
